src/cluster/placement/algo/mirrored.go

```diff
@@ -289,17 +355,18 @@ func allInstancesInState(
 	p placement.Placement,
 	forEachShardFn func(s shard.Shard) bool,
 ) bool {
-	for _, instance := range p.Instances() {
-		if !instanceCheck(instance, forEachShardFn) {
-			continue
+	for id := range instanceIDs {
```

Comment:
So I feel like this function now has very different semantics than it did:

previously: return true iff the instances in `instanceIDs` are the only ones in the placement which pass `forEachShardFn`
new: check that each instance in `instanceIDs` passes `forEachShardFn` (ignore all other instances).

Assuming that's a correct summary, what's the reasoning behind the change? Does it cause issues? One thing we might want to consider: this code path gets hit several times (afaict) in a lot of our assertions in tests, so changing this method here has the side effect of changing our assertions. It might be worth temporarily splitting out the functions we use for the assertions from this function just to make sure we're not causing unintentional regressions there, i.e.:

```
// use old behavior in tests (naming is bad sorry; it's a weird function).
func onlyTheseInstancesInState()

// prod code uses modified `allInstancesInState`
```

Reply:
The reasoning is to allow non-overlapping replaces. This change of semantics, as coded, also applies to other types of placement changes (add/remove) to be done "in parallel" as long as they are non-overlapping. We could, of course, relax scope this change of semantics only for replaces and not any other types of placement changes. To keep the code simple I'd suggest to apply it for all because node replace is the most common type of placement change during the lifetime of a service, whereas add/remove is rather rare. In practice it means it allows add/remove/replace during a pending add/remove/replace if shards are not overlapping.

Only one of existing tests have failed due to change of this semantics: `TestMirrorAddMultiplePairs` and needs to be updated, all other tests pass without changing their assertions.

Reply:
An argument towards scoping this new semantics only to replaces is to disallow add/remove during pending replace or during another another add/remove. Add/remove is presumably always initiated by human, is not an emergency and could wait for the sake of minimizing risks of chain of undesired placement changes.

Reply:
As we concluded, it is best to keep old semantics for add/remove and use new semantics only for replace. I will extract the logic of those checks into separate structs that will look something like this:
```
globalChecker.allInitializing(...)
vs.
exactChecker.allInitializing(...)
```